encoding/custom/common_codec/common.go

```diff
@@ -0,0 +1,113 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2022 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package common_codec
+
+import (
+	"fmt"
+	"io"
+)
+
+type LengthyWriter struct {
+	w      io.Writer
+	length int
+}
+
+func NewLengthyWriter(w io.Writer) LengthyWriter {
+	return LengthyWriter{w: w}
+}
+
+func (l *LengthyWriter) Write(p []byte) (n int, err error) {
+	n, err = l.w.Write(p)
+	l.length += n
+	return
+}
+
+func (l *LengthyWriter) Len() int {
+	return l.length
+}
+
+type LocatedReader struct {
+	r        io.Reader
+	location int
+}
+
+func NewLocatedReader(r io.Reader) LocatedReader {
+	return LocatedReader{r: r}
+}
+
+func (l *LocatedReader) Read(p []byte) (n int, err error) {
+	n, err = l.r.Read(p)
+	l.location += n
+	return
+}
+
+func (l *LocatedReader) Location() int {
+	return l.location
+}
+
+func Concat(deep ...[]byte) []byte {
+	length := 0
+	for _, b := range deep {
+		length += len(b)
+	}
+
+	flat := make([]byte, 0, length)
+	for _, b := range deep {
+		flat = append(flat, b...)
+	}
+
+	return flat
+}
+
+type EncodedBool byte
+
+const (
+	EncodedBoolUnknown EncodedBool = iota
+	EncodedBoolFalse
+	EncodedBoolTrue
+)
+
+func EncodeBool(w io.Writer, boolean bool) (err error) {
+	b := EncodedBoolFalse
+	if boolean {
+		b = EncodedBoolTrue
+	}
+
+	_, err = w.Write([]byte{byte(b)})
+	return
+}
+
+func DecodeBool(r io.Reader) (boolean bool, err error) {
+	b := make([]byte, 1)
+	_, err = r.Read(b)
+	if err != nil {
+		return
+	}
+
+	switch EncodedBool(b[0]) {
+	case EncodedBoolFalse:
+		boolean = false
+	case EncodedBoolTrue:
+		boolean = true
+	default:
+		err = fmt.Errorf("invalid boolean value: %d", b[0])
```

Comment:
Could you please add test cases for this uncovered code reported by codecov and other error / unhappy code paths?

Reply:
Punting this for later, pre-merge into master: https://github.com/onflow/cadence/issues/1939